### ISSUE TITLE
Option 3 randomly decides to stop long before reaching actual last page

### ISSUE THREAD

**Issue by AgentThirteen** [Nandaka/PixivUtil2#636]
### Prerequisites

* [x] Did you read FAQ section?
* [x] Did you test with the latest [releases](https://github.com/Nandaka/PixivUtil2/releases "latest releases") or [commit](https://github.com/Nandaka/PixivUtil2/commits/master "latest commit") ?

### Description

Option 3 may decide to stop long before reaching the actual last page, sometimes stopping as early as page 1 as the last page. This happens both with small (<1k results) and large (up to 50k results) queries but odds are higher with a larger count since more pages get processed. Works getting pushed in the meantime may also affect this.

### Steps to Reproduce

1. Option 3 (download by tags), any query - if using the binary do not pass x when done, if using a command-line do not pass an exit command to the command-line when done
2. Don't set a last page or set it high enough for the testing period
3. Wait until the process stops prematurely, results so far yielded anything between page 1 and page 281 when the actual end page was over 800 so this could take a very long time if tested like this; it has also happened several times in a row for the same query with different responses

**Expected behavior:**
Script reaches the last work before ending.

**Actual behavior:**
Script ends prematurely and has to be monitored, returning one of the following:
- Last page reached (most frequent)
- No more images + last page reached
- Getting duplicate results, stopping (rare even for a large query)

### Versions

Latest Windows binary release `20200125`

(Cannot deploy a Python 3 environment for a while and all Python 2 releases - source and binaries - will fail due to cookie handling after the site change a few days ago and #505 seems to be limited to Python 3 libraries)

### Additional info

This issue might be linked to a #550-related change - did not change versions until the latest site revamp.

Not using a proxy or VPN. Not blocked in any way, browsers or otherwise. No indications of a server-side action triggering this.

The script is not blocked when this happens, any other running instance will still work and affected instances can resume (reloading the `config.ini` file) once new arguments with the next page are passed without closing the script, as if nothing unusual happened.

Dumps don't really provide useful information for this issue, no errors, the script just ends too early with a quite unexpected last page reached response. I couldn't really see a pattern in any of the binary outputs at the time of stopping.

Thanks again for your work Nandaka and everyone and sorry for the lack of issue activity and feedback.

edits: fixed wrong cross-referenced issue number, removed potentially dangerous workarounds

**Comment by Nandaka** [comment]
what is the tag you are using? It is using their API (e.g. https://www.pixiv.net/ajax/search/artworks/vocaloid?word=vocaloid&order=date_d&mode=all&p=1&s_mode=s_tag&type=all),  so if they return empty/no image, then it will stop. The url also recorded in the log file.

In very rare case or if you are using generic tag, more than 60 images might be uploaded in the server at the same time you getting the previous page.

**Comment by AgentThirteen** [comment]
Thank you very much for the very quick reply Nandaka. I am aware it uses the ajax API response and that might be what happens if the timing is unlucky for active queries (frequent new posts).

`In very rare case or if you are using generic tag, more than 60 images might be uploaded in the server at the same time you getting the previous page.`
That would be the most plausible explanation for the affected queries. Retrying the same page won't fix it until another work has been pushed so it is likely workcount-related. Retrying immediately from the next page will work fine until the next occurence (and retrying from the previous page will fetch it fine then stop again immediately after that page).

Upon further inspection, this happens to three partial match queries (multiple tags with boolean routines and exclusions) all yielding over 5k results and fairly active. Another query like these has not had this issue so far but it browses backwards (old posts first).

I also checked the stopping URLs both from full dump and the latest binary itself immediately to make sure no works were pushed in the meantime - binary output randomly showed 20, 30, 58 or the regular 59 works (no occurence of 60 to my knowledge) getting parsed at the time of stopping and while one of the works before/after was an ugoira, others were standard pictures or manga and still caused this. Same result for the previous and next pages, nothing in common that could have triggered this.

I don't know exactly when this started to happen in the project timeline since the `20191105` command-line version was working fine with the quickfix back then (#550).

**Comment by Ksdpm** [comment]
It seems to be happening to me as well. The 'time period' search seems to be capped at 1 year after several trials. I not very good at non-scientific coding so I just loop the date, i.e. 1 year each time, for the search.

**Comment by AgentThirteen** [comment]
Thanks for the input @Ksdpm and sorry for the late follow-up reply. This one is not related to the 1-year limitation and will happen completely randomly with increasing odds for very large queries, high activity searches or both and still affects current versions.

The current workaround that was suggested is to restart the script with the following page as the first page to process (using the same page will always return `No more images` even though there are more), which will return `No more images` and exit in a clean manner if that really was the last page or resume where it left off if there is more to fetch.

Case in point, low-result, very active query stopped on page 3 out of 24, extremely large, extremely active query stopped on pages 275 and 841 out of 950 and large, very active query stopped on page 120 out of 243. Conditions unclear as before.

Premium users can test whether they are affected but it is not recommended at all to test extremely large queries that have the potential of draining a lot of server resources with a premium account and sensitive information. `blacklist_members` and `blacklist_tags` were suspected at some point as well as a specific result count but that last February post shows they were not involved and it may be some kind of interstitial element or a new work getting pushed (which should still make things fail much faster for very large and extremely active queries).

Make sure you also save logs in different folders or files to troubleshoot with parallel processes or your dump log will not mean much as queries will collide and it becomes hard to tell why a specific one stopped and what it was trying to do when it happened (pausing the console with a decent buffer size usually is enough to check the output for the last 2 pages as long as not many files were downloaded if any at all).

**Comment by AgentThirteen** [comment]
This took a little while but I think I have found the culprit. This happens when a work has been **deleted** and leaves behind an empty container that does not have any HTML element identifier in the DOM, and does not show up in the JSON response either.

Not a `<li>` element and no container of any kind then closes the entire indexed list with `</ul>` so the JSON response stops at zero-indexed 58 here, effectively causing the script to think there are no more images and that this is the last page. This seems to fix itself when it is pushed to the next page so this looks like a pixiv formatting bug when something that is listed last gets deleted.

![deleted](https://user-images.githubusercontent.com/19692328/83407694-920dd200-a411-11ea-9539-7bffc065f25f.png)

Associated JSON response with just the important things:
```
{"error":false,
"body":
 {"illustManga":
  {"data":
   {0: {...}
   {58: {...}
   {total: *
  }
 }
}
```

Maybe trying to force fetch the next page when hitting exactly 59 results was the best approach after all. I hope this logging output will be of help.

**Comment by Nandaka** [comment]
Weird, current logic already count the ad container, do you have the full json?
https://github.com/Nandaka/PixivUtil2/blob/master/PixivTags.py#L55
https://github.com/Nandaka/PixivUtil2/blob/master/PixivTags.py#L76

**Comment by AgentThirteen** [comment]
Thank you again for the quick reply Nandaka.

Sorry, I meant 58 results in the console when the script halts. That's 59 including the ad container (which was a selectable HTML element on the page and doesn't cause issues). On a page with the container but no ghost element, the JSON response is indexed from 0 to 59 as expected.

Unfortunately I don't have the full JSON at hand but since it was from 0 to 58, this probably includes the container and not the ghost element. In order to reproduce it, it is crucial to look at the page that caused the issue before any other work gets pushed (in a browser's HTML output) as this doesn't appear to be a true element at all.

What I still don't understand is how a deleted work (at least that's how deleted works appeared in the past, though they used to be actual elements) that is at the end of a page getting pushed to the next one by a new one fixes the issue. When this happens, the affected page has the expected 60 count (59+ad) but the next page also has 60, the deleted work "ghost" frame doesn't seem to get pushed to the beginning, it vanishes completely and everything goes back to normal...

If I run into this again and manage to catch it at the right time, I will try and get back will the full JSON response and content, sorry about this. It would also be very kind if someone could please confirm this happens exactly in 58-result page cases as this greatly narrows down safety checks on the last page being an actual last page. I guess running again from an extra page each time the script comes to an end is unusual enough not to be a performance concern but it's probably better not to do such a thing if we have a simple way to guarantee there are more results to the query.

**Comment by AgentThirteen** [comment]
This happened several times today and I managed to get a capture.
Raw log has a lot to sanitize so I will provide relevant information in readable format.

Here is the sanitized JSON as a screenshot and highlights so you do not have to waste time on it. :smile:

- `data` length becomes 59 here, 58 work results plus ad but doesn't explain https://github.com/Nandaka/PixivUtil2/issues/636#issuecomment-590062955 :confused:
- `isAdContainer` was randomized at index 40 (41st array element), not before or after the ghost frame, only important part is it is there and counted
- total results in the query was way less than 59k so value was removed manually
- ghost element as stated does not appear at all, not in the DOM, not in the JSON response

![sanitized_json](https://user-images.githubusercontent.com/19692328/83639477-092a9e00-a5ab-11ea-9837-bb5c6818e5e6.png)
I still cannot really figure out what causes this, only that it started after the logic was changed from 48 to 60 to match the result change. I completely forgot about the previous comment where the residual frame from a deleted work - this is the only thing I am almost sure of about this ghost frame - may have been anywhere on the page.

The workaround that I still use is to run **the next page** when ending with less than 59 actual works but this will only work in regular chronological search mode. The next page is guaranteed to return no more images if it is the last page and also guaranteed to try and fetch more including the result that was pushed back by the ghost frame(s) so no work is lost during the search, only maybe performance...?
